iree/compiler/Codegen/Common/ForOpCanonicalizationPass.cpp

```diff
@@ -50,86 +94,184 @@ namespace {
+        forOp.getLoc(), forOp.lowerBound(), forOp.upperBound(), forOp.step(),
+        initValues);
+    transferBody(forOp.getBody(), newLoop.getBody(), yieldValues, rewriter);
+
+    // Replace the results with the new loop's results.
+    SmallVector<Value, 8> newResults(newLoop.getResults().begin(),
                                      newLoop.getResults().end());
-    for (auto en : llvm::enumerate(iteratorFolded)) {
+    for (const auto& it : llvm::enumerate(foldedCarriedValueIndices)) {
+      // The current iterator argument involves shape ops that can be cancelled
+      // across the region boundary. This means pushing the original def to
+      // after the region. So clone it outside and take the new result as
+      // operands.
       BlockAndValueMapping mapping;
-      mapping.map(returnValues[en.value()], newLoop.getResult(en.value()));
-      repResults[en.index()] =
-          rewriter.clone(*resultOps[en.index()], mapping)->getResult(0);
+      mapping.map(yieldValues[it.value()], newLoop.getResult(it.value()));
+      newResults[it.index()] =
+          rewriter.clone(*resultOps[it.index()], mapping)->getResult(0);
+
       Operation* oldOp =
-          newLoop.getRegionIterArgs()[en.index()].use_begin()->getOwner();
-      SmallVector<Value, 1> arg(1, newLoop.getRegionIterArgs()[en.index()]);
+          newLoop.getRegionIterArgs()[it.index()].use_begin()->getOwner();
+      SmallVector<Value, 1> arg(1, newLoop.getRegionIterArgs()[it.index()]);
       oldOp->replaceAllUsesWith(arg);
     }
-    rewriter.replaceOp(forOp, repResults);
+    rewriter.replaceOp(forOp, newResults);
+    return success();
+  }
+};
+
+/// Fold reciprocal ops across IfOp boundary. This is similar to the above
+/// pattern for ForOps.
+///
+/// This pass does this kind of transformation:
+/// ```
+/// %0 = scf.if %cond -> (vector<1x4xf32>) {
+///   %1 = scf.broadcast %scalar : f32 into vector<1x4xf32>
+/// } else {
+///   scf.yield %zero : vector<1x4xf32>
+/// }
+/// %2 = vector.extract %0[0, 0] : vector<1x4xf32>
+/// ```
+/// ->
+/// ```
+/// %0 = scf.if %cond -> (f32) {
+///   scf.yield %scalar : f32
+/// } else {
+///   scf.yield %zero : f32
+/// }
+/// ```
+struct CanoncalizeIfReturnedValueShape final
+    : public OpRewritePattern<scf::IfOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::IfOp ifOp,
+                                PatternRewriter& rewriter) const override {
+    auto resultTypes = llvm::to_vector<8>(ifOp.getResultTypes());
+    if (resultTypes.empty()) return failure();
+
+    // Require all yield values in the "else" branch to be splat constant for
+    // now. Merging different cases between the "then" and "else" branches can
+    // be complicated; leave that to only when needed.
+    if (ifOp.elseBlock()) {
+      SplatElementsAttr splat;
+      for (Value v : ifOp.elseYield().getOperands()) {
+        if (!matchPattern(v, m_Constant(&splat))) return failure();
+      }
+    }
```

Comment:
That seems very ad hoc. I think you can make the general case work by just cloning the extract in the else case. This should simplify the logic too.

Reply:
Good point! Done. Thanks for the suggestion!